Jake Ladder step3

ChocOZerO opened the pull request:
split 후에 문자 처리하는게 String.SubSequence 로 받아서 처리하는게 맞는지 모르겠어요.
String이랑 달라서 중간중간 형변환에 고생한것 같습니다.

ChocOZerO committed:
[Refactoring]구조체를 활용하여 사다리게임 구현

기존에 함수로만 되어있던 로직개선
LadderPlayer와 LadderGame을 활용해 로직구현
InputView와 ResultView를 활용해 입출력 구분

ChocOZerO committed:
[Update]플레이어이름 5글자로 통일하여 출력

ChocOZerO committed:
[Add]플레이어 이름 입력받을때 validation check

ChocOZerO committed:
[Refactoring]struct별로 파일 분할

ChocOZerO committed:
[Update]플레이어 이름 포맷팅 padding을 이용해 개선

godrm reviewed (commented):
커밋 메시지 머릿말 표시는 좋네요. 
- Main에 있는 함수는 불필요해보입니다. 아래 내용을 참고해서 Main 부분의 코드를 줄여보세요. 
입력을 다시 받는 while문 자체도 InputView에서 처리하도록 개선하세요. 
- makeLadderLayer에 ladderHeight를 또 넘겨야 하나요? 생성할 때 이미 넘긴 값을 중복해서 처리하는 것 같습니다.

ChocOZerO committed:
[Refactoring]InputView가 입력로직 전담하도록 개선

반복문등의 validation과 입력에 관한 모든 로직을 InputView로 이동
main에서는 새로운 게임을 알리는 setNewGame을 이용해 호출만 함
새로운 타입인 LadderGameSetter라는 튜플을 이용
입력값을 받아와서 메인에서는 리턴된 튜플값으로 세팅

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,76 +8,23 @@
+    // 사다리 게임 필수 값 지정
+    let playerNames : [String.SubSequence] = ladderGameSetter.playerNames
+    let ladderHeight : Int = ladderGameSetter.ladderHeight
+    var ladderGame : LadderGame = LadderGame(height: ladderHeight, names: [], ladder: [])
```
names 와 ladder 가 불필요한 매개변수라면 default 값을 지정하는 init 함수를 만드는게 더 좋습니다.
그리고 playerNames를 그대로 names에 넘길 방법은 없을까요? 
초기화 init 함수에 필요한 값들을 다 넘기고 시작할 수 있는게 가장 좋은 방법입니다.

ChocOZerO replied:
initializer 추가했습니다.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    var height : Int = 0
+    var names : [LadderPlayer] = []
+    var ladder : Array<Array<String>> = []
```
var height 와 var names 를 private 으로 변경하고 get 만 할 수 있도록 개선하세요.
객체 속성은 객체 외부에서 바꿀 수 없도록 감추고 인터페이스를 통해서만 변경하도록 만들어야 합니다. 
var ladder도 외부에서 참조한다면 동일하게 변경하세요. 참조하지 않는다면 private으로 지정하세요.

ChocOZerO replied:
변수는 전부 private로 바꾸고 get을 활용하긴했는데 swift에서는 getNames() 이런식으로 안쓰는거 같네요. 습관처럼 getNames라고 이름을 붙이긴 했는데 어떤식으로 활용하는게 좋을까요? 현재 제 방식에서 이상한 부분 있으면 알려주세요.
현재 LadderPlayer랑 LadderGame둘다 변경했습니다.

ChocOZerO committed:
[Update]플레이어이름 5글자 자동 세팅 개선

ChocOZerO committed:
[Refactoring]Add Initializer, getter on struct

LadderGame, LadderPlayer 은닉화
Initializer추가
변수들을 private로 변경우 getter를 통해 접근하도록 변경

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,76 +8,21 @@
+    // 사다리 게임 인스턴스 생성
+    var ladderGame : LadderGame = LadderGame(height: ladderGameSetter.ladderHeight, names: ladderGameSetter.playerNames)
+    // 입력된 값을 이용해 사다리게임 만들기
+    ladderGame.makeLadderLayer(playerNames: ladderGameSetter.playerNames)
```
이미 생성할때 names를 넘기는데, makeLadderLayer() 함수 playerNames 매개변수가 동일한 게 필요없지 않나요? 확인해주세요.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,66 @@
+            return self.names
+        }
+    }
+    var getLadder : Array<Array<String>> {
```
var getNames와 var getLadder는 getter 인터페이스가 아니라 get- 이름이 붙은 새로운 속성을 만들었네요. 
스위프트의 private (get) 설정 방식을 활용하거나 메소드로 만드는게 좋습니다.

ChocOZerO replied:
그렇다면 변화가 없는 let 변수 같은 경우엔 initializer이후에 값 변화가 없을텐데 private자체가 필요없고 let으로 선언해도 괜찮을까요??
예를들면 제 경우에 LadderGame의 height라던가 LadderPlayer의 name 변수를 상수(let)으로 바꾸고 굳이 private(set)을 활용하지 않아도 되는지 궁금합니다.

godrm replied:
초기값 이후에 바뀌지 않으면 당연히 let 으로 하는게 더 좋습니다.

ChocOZerO committed:
[Update]getter 인터페이스를 활용한 방식으로 수정

private(set), let의 활용법 습득

ChocOZerO committed:
[Refactoring]ResultView 전체 출력메소드 하나만 노출

사다리와 플레이어의 프린트를 각각 호출했던 기존 방식 변경
전체 결과를 외부에서 호출하면 한번에 사다리와 플레이어 출력

ChocOZerO committed:
[Update]LadderGame의 height private 선언

godrm commented:
앞으로도 객체단위로 역할과 책임을 분리할 때 private 변수/함수를 만들어야 합니다.
머지하겠습니다 👍
